dotCMS/src/main/java/com/dotcms/business/SystemTableFactoryImpl.java

```diff
@@ -0,0 +1,130 @@
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+
+/**
+ * Default implementation of {@link SystemTableFactory}
+ */
+public class SystemTableFactoryImpl extends SystemTableFactory {
+
+    private final SystemCache systemCache;
+
+    private final static String VALUE_404 = "DOT_SYSTEM_TABLE_404";
+
+    public SystemTableFactoryImpl () {
+
+        this.systemCache = CacheLocator.getSystemCache();
+    }
+    @Override
+    protected Optional<String> find(final String key) throws DotDataException {
+
+        Object value = null;
+        if(UtilMethods.isSet(key)) {
+
+            value = this.systemCache.get(key);
+            if(Objects.isNull(value)) {
+
+                final List<Map<String, Object>> result = new DotConnect()
+                        .setSQL(" SELECT * FROM system_table WHERE key = ? ")
+                        .addParam(key)
+                        .loadObjectResults();
+
+                value = result.isEmpty() ? null : result.get(0).get("value");
+
+                if(Objects.nonNull(value)) {
+                    this.systemCache.put(key, value);
+                } else {
+                    this.systemCache.put(key, VALUE_404);
+                    value = VALUE_404;
+                }
+            }
+        }
+
+        return  Objects.isNull(value) || VALUE_404.equals(value)?
+                Optional.empty():
+                Optional.ofNullable(toString(value));
+    }
+
+    @Override
+    protected Map<String, String> findAll() throws DotDataException {
+
+        final Map<String, String> records = new LinkedHashMap<>();
+
+        final List<Map<String, Object>> result = new DotConnect()
+                .setSQL(" SELECT * FROM system_table")
+                .loadObjectResults();
+
+        for (final Map<String, Object> recordMap : result) {
+
+            final Object key = recordMap.get("key");
+            final Object value = recordMap.get("value");
+            if (Objects.nonNull(key) && Objects.nonNull(value)) {
+                records.put(key.toString(), toString(value.toString()));
+                this.systemCache.put(key.toString(), value);
+            }
+        }
+
+        return records;
+    }
+
+    @Override
+    protected void saveOrUpdate(final String key, final String value) throws DotDataException {
+
+        if (Objects.nonNull(key) && Objects.nonNull(value)) {
+
+            new DotConnect()
+                    .setSQL( //upsert
+                            "INSERT INTO system_table (value, key) values(?,?)" +
+                                    "ON CONFLICT(key) DO UPDATE SET value = ?")
+                    .addParam(value)
+                    .addParam(key)
+                    .addParam(value)
+                    .loadResult();
+
+            this.systemCache.remove(key);
+        } else {
+
+            throw new DotDataException("The key and value should not be null");
```

Comment:
Do we really have to throw an exception when this happens? Can't we just ignore it?

Reply:
Agree, if the key is null just return and maybe if the value is null, we just delete the key

Reply:
I think we do